core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@junrao No, I do not mean that sending RPC requests dominate the cost. The answer to question (1) is yes and the answer to question (2) is no. What I mean is that it will take much longer time for broker to reach StopReplicaResponse sent by controller, if controller sends 3000 requests instead of 3 requests (not counting StopReplicaResponse). The time for broker to reach StopReplicaResponse in the channel between itself and controller will dominate the cost.

Using the same example illustrated above, the channel from controller to broker will have 3000 requests + 1 StopReplicaResponse. In order to reach this StopReplicaResponse, broker would need to take 3000 requests out of the channel first. Say the broker has 9 other channels (e.g. with client) that are also actively sending request. Broker would need to process `3000*10` requests in total before reaching that StopReplicaResponse. This time can be much longer than 3000 ZK writes.